lens/fetcher.go

```diff
@@ -0,0 +1,400 @@
+// Copyright 2023 Democratized Data Foundation
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package lens
+
+import (
+	"context"
+	"reflect"
+
+	"github.com/fxamacker/cbor/v2"
+
+	"github.com/sourcenetwork/defradb/client"
+	"github.com/sourcenetwork/defradb/client/request"
+	"github.com/sourcenetwork/defradb/core"
+	"github.com/sourcenetwork/defradb/datastore"
+	"github.com/sourcenetwork/defradb/db/fetcher"
+	"github.com/sourcenetwork/defradb/planner/mapper"
+)
+
+// todo: The code in here can be significantly simplified with:
+// https://github.com/sourcenetwork/defradb/issues/1589
+
+type lensedFetcher struct {
+	source   fetcher.Fetcher
+	registry client.LensRegistry
+	lens     Lens
+
+	txn datastore.Txn
+
+	col *client.CollectionDescription
+	// Cache the fieldDescriptions mapped by name to allow for cheaper access within the fetcher loop
+	fieldDescriptionsByName map[string]client.FieldDescription
+
+	targetVersionID string
+}
+
+var _ fetcher.Fetcher = (*lensedFetcher)(nil)
+
+// NewFetcher returns a new fetcher that will migrate any documents from the given
+// source Fetcher as they are are yielded.
+func NewFetcher(source fetcher.Fetcher, registry client.LensRegistry) fetcher.Fetcher {
+	return &lensedFetcher{
+		source:   source,
+		registry: registry,
+	}
+}
+
+func (f *lensedFetcher) Init(
+	col *client.CollectionDescription,
+	fields []client.FieldDescription,
+	filter *mapper.Filter,
+	docmapper *core.DocumentMapping,
+	reverse bool,
+	showDeleted bool,
+) error {
+	f.col = col
+
+	f.fieldDescriptionsByName = make(map[string]client.FieldDescription, len(col.Schema.Fields))
+	// Add cache the field descriptions in reverse, allowing smaller-index fields to overwrite any later
+	// ones.  This should never really happen here, but it ensures the result is consistent with col.GetField
+	// which returns the first one it finds with a matching name.
+	for i := len(col.Schema.Fields) - 1; i >= 0; i-- {
+		field := col.Schema.Fields[i]
+		f.fieldDescriptionsByName[field.Name] = field
+	}
+
+	f.targetVersionID = col.Schema.VersionID
+	return f.source.Init(col, fields, filter, docmapper, reverse, showDeleted)
+}
+
+func (f *lensedFetcher) Start(ctx context.Context, txn datastore.Txn, spans core.Spans) error {
+	history, err := getTargetedSchemaHistory(ctx, txn, f.registry.Config(), f.col.Schema.SchemaID, f.col.Schema.VersionID)
```

Comment:
Could add an early return, but it would only save a few nanoseconds I think. `lens` initialisation involves pretty much nothing. I'd rather have simple code.

Reply:
I think its more about clear intent, if there is nothing todo, then we shouldnt be allocating and instanciating objects that are doing nothing. It adds to the complexity of the reader who has to read and grok a lot more lines spread out elsewhere to fully determine/gurantee that if no history exists, then there is no overhead to the call, which requires understanding `getHistory` more extensively, which is a non-trivial function.

A simple `if` block signals to the reader the empty condition is satisfied, and theres no need to build out a lens pipeline.

Not so much saving a few nanoseconds.

Reply:
This has been handled in Init due to the fetcher-fields issue.